File: JS/PassChecker/PasswordChecker.test.js
import { checkPassword } from './PasswordChecker.js'
import {describe, test} from 'node:test'
import * as assert from 'node:assert'

describe('checkPassword test suite', () => {
    test('should throw an error if the password is not a string', () => {
        assert.throws(() => checkPassword(123), Error('Password must be a string'))
    });
    test('should throw an error if the password is less than 8 characters long', () => {
        assert.throws(() => checkPassword('abc'), Error('Password must be at least 8 characters long'))
    });
    test('should throw an error if the password is more than 20 characters long', () => {
        assert.throws(() => checkPassword('abcdefghijabcdefghijabcdefghij'), Error
        ('Password must be at most 20 characters long'))
    });
    test('should throw an error if the password does not contain at least one lowercase letter', () => {
        assert.throws(() => checkPassword('ABC12345'), Error('Password must contain at least one lowercase letter'))
    });
    test('should throw an error if the password does not contain at least one uppercase letter', () => {
        assert.throws(() => checkPassword('abc12345'), Error('Password must contain at least one uppercase letter'))
    });
    test('should throw an error if the password does not contain at least one digit', () => {
        assert.throws(() => checkPassword('Abcdefgh'), Error('Password must contain at least one digit'))
    });
    test('should throw an error if the password does not contain at least one special character', () => {
        assert.throws(() => checkPassword('Abcdefgh1'), Error('Password must contain at least one special character'))
    });

    test('should throw error if the password contains the ! character', ()=>{

        assert.throws(() => checkPassword('Abcdefgh1!'), Error('Password must contain at least one special character'))
    })

    test('should not throw an error if the password meets all the criteria', () => {
        assert.doesNotThrow(() => checkPassword('Abcdefgh1?'))
    });


})

File: Java/my-app/src/test/java/com/barosanu/PasswordCheckerTest.java
package com.barosanu;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.BeforeEach;

public class PasswordCheckerTest {

    private PasswordChecker checker;

    @BeforeEach
    public void setUp() {
        checker = new PasswordChecker();
    }

    @Test
    public void testValidPassword() {
        assertTrue(checker.checkPassword("Abcdefg1!"));
    }

    @Test
    public void testShortPassword() {
        assertFalse(checker.checkPassword("Abc1!"));
    }

    @Test
    public void testLongPassword() {
        assertFalse(checker.checkPassword("Abcdefg123456789012345!"));
    }

    @Test
    public void testNoLowerCaseLetter() {
        assertFalse(checker.checkPassword("ABCDEFG1!"));
    }

    @Test
    public void testNoUpperCaseLetter() {
        assertFalse(checker.checkPassword("abcdefg1!"));
    }

    @Test
    public void testNoDigit() {
        assertFalse(checker.checkPassword("Abcdefg!"));
    }

    @Test
    public void testNoSpecialCharacter() {
        assertFalse(checker.checkPassword("Abcdefg1"));
    }

    @Test
    public void testPasswordTooShort() {
        Exception exception = assertThrows(Exception.class, () -> {
            checker.checkPasswordAndThrowReason("Abc1!");
        });
        assertEquals("Password must be at least 8 characters long", exception.getMessage());
    }

    @Test
    public void testPasswordTooLong() {
        assertThrows(Exception.class, () -> {
            checker.checkPasswordAndThrowReason("Abcdefg1234567890123456!");
        });
    }

    @Test
    public void testNoLowerCaseLetterWithReason() {
        Exception exception = assertThrows(Exception.class, () -> {
            checker.checkPasswordAndThrowReason("ABCDEFG1!");
        });
        assertEquals("Password must contain at least one lowercase letter", exception.getMessage());
    }

    @Test
    public void testNoUpperCaseLetterWithReason() {
        Exception exception = assertThrows(Exception.class, () -> {
            checker.checkPasswordAndThrowReason("abcdefg1!");
        });
        assertEquals("Password must contain at least one uppercase letter", exception.getMessage());
    }

    @Test
    public void testNoDigitWithReason() {
        Exception exception = assertThrows(Exception.class, () -> {
            checker.checkPasswordAndThrowReason("Abcdefg!");
        });
        assertEquals("Password must contain at least one digit", exception.getMessage());
    }

    @Test
    public void testNoSpecialCharacterWithReason() {
        Exception exception = assertThrows(Exception.class, () -> {
            checker.checkPasswordAndThrowReason("Abcdefg1");
        });
        assertEquals("Password must contain at least one special character", exception.getMessage());
    }

    @Test
    public void testValidPasswordWithReason() {
        assertDoesNotThrow(() -> {
            checker.checkPasswordAndThrowReason("Abcdefg1!");
        });
    }
    
}
